### PR TITLE
.Net: Fix #6030 - Mitigating Prompt Injection in Liquid Templates

### LINKED ISSUE #6030: .Net Implement AllowUnsafeContent to prompty's input variables
See comments
https://github.com/microsoft/semantic-kernel/pull/5961#discussion_r1579136407

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
### Motivation and Context

#6030 
### Description
~In this implementation, the `Ġ` will be reserved in liquid template which is used to replace `:` in all input variables when unsafe content is not allowed.~

~The encoding process for input variables when unsafe content is not allowed is~
~- replace all `:` to `Ġ` // this is the extra step comparing with HandlerBar Template~
~- Encode xml using `HttpUtility.HtmlEncode`~

~The decoding process is~
~- replace all `Ġ` to `:`~

This PR introduces a new process to mitigate potential prompt injection attacks from input variables when using liquid templates. Here's a breakdown of the steps:

Before rendering, each input variable undergoes a transformation: all occurrences of `:`are replaced with `&#58;`. This ensures that message tags like `system:`, `user:`, or `assistant:` are not present if `AllowUnsafeContent` is set to `false`. No replacement occurs if `AllowUnsafeContent` is `true`.
After rendering, each message content is processed based on the `AllowUnsafeContent` setting. If it's `false`, all `&#58;` instances are reverted back to `:`, followed by calling `html_encode` on each message content. If `AllowUnsafeContent` is `true`, only `html_encode` is called. This additional encoding step is necessary because `ChatPromptParser` always decodes XML message content, requiring the liquid template to undergo an extra encoding step to ensure the rendered content matches the original before rendering.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs [hunk 1]
```diff
@@ -65,13 +71,106 @@ public Task<string> RenderAsync(Kernel kernel, KernelArguments? arguments = null
...
+    /// <param name="template"></param>
+    /// <returns></returns>
+    private string PreProcessTemplate(string template)
+    {
+        if (this._allowUnsafeContent)
+        {
+            return template;
+        }
+
+        if (template.Contains(ReservedChar))
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:100]: What if my prompt template asks LLM to give me the history of the Ġ character?

File: dotnet/src/Extensions/PromptTemplates.Liquid.UnitTests/LiquidTemplateTest.ItRendersAndCanBeParsedAsync.verified.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+﻿system:This is the system message
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/PromptTemplates.Liquid.UnitTests/LiquidTemplateTest.ItRendersAndCanBeParsedAsync.verified.txt:1]: Where are the *.verified.txt files used? Are unit tests consuming them, or perhaps generating them? If it's the former, why are they not part of the project? If it's the latter, why is this necessary?

**Comment by stephentoub** [comment]
> In this implementation, the Ġ will be reserved in liquid template which is used to replace : in all input variables when unsafe content is not allowed.

Why is that necessary? And why is that ok? Why are we in a position to effectively ban "Ġ"?

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs [hunk 3]
```diff
@@ -65,13 +71,106 @@ public Task<string> RenderAsync(Kernel kernel, KernelArguments? arguments = null
...
+            return text;
+        }
+
+        return text.Replace(ReservedChar, ColonChar);
+    }
+
+    /// <summary>
+    /// Gets the variables for the prompt template, including setting any default values from the prompt config.
+    /// </summary>
+    private KernelArguments GetVariables(Kernel _, KernelArguments? arguments)
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:122]: Sure

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplateFactory.cs [hunk 4]
```diff
@@ -15,12 +15,24 @@ public sealed class LiquidPromptTemplateFactory : IPromptTemplateFactory
...
+    /// <summary>
+    /// Gets or sets a value indicating whether to allow unsafe content.
+    /// </summary>
+    /// <remarks>
+    /// The default is false.
+    /// When set to true then all input content added to templates is treated as safe content and will not be HTML encoded.
+    /// For prompts which are being used with a chat completion service this should be set to false to protect against prompt injection attacks.
+    /// When using other AI services e.g. Text-To-Image this can be set to true to allow for more complex prompts.
+    /// </remarks>
+    [Experimental("SKEXP0001")]
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplateFactory.cs:27]: Thanks will do

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs [hunk 5]
```diff
@@ -30,9 +35,10 @@ public Task<string> RenderAsync(Kernel kernel, KernelArguments? arguments = null
         Verify.NotNull(kernel);
 
         var template = this._config.Template;
+        template = this.PreProcessTemplate(template);
         var liquidTemplate = Template.ParseLiquid(template);
-        var nonEmptyArguments = arguments?.Where(x => x.Value is not null).ToDictionary(x => x.Key, x => x.Value!);
-        var renderedResult = liquidTemplate.Render(nonEmptyArguments);
+        arguments = this.GetVariables(kernel, arguments);
+        var renderedResult = liquidTemplate.Render(arguments.ToDictionary(x => x.Key, x => x.Value));
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:41]: Maybe? Let me try out

File: .gitignore [hunk 6]
```diff
@@ -490,3 +490,6 @@ swa-cli.config.json
 
 # python devcontainer
 /python/.devcontainer/*
+
+# dotnet Verify
+**/*.received.txt
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply .gitignore:495]: TLDR:
This is to prevent `*.received.txt` from being checked into the repo.

A bit more explanation:
the intention is to avoid committing *.received.txt files to the repository. These files are generated during testing by the Verify function and are compared with corresponding *.verified.txt files. If no differences are found, the test passes and the *.received.txt file is deleted. If there are differences, the test fails, and the *.received.txt file remains. In both cases, received.txt files are temporary and should not be included in the repository since they are only used for comparison purposes

File: dotnet/src/Extensions/PromptTemplates.Liquid.UnitTests/LiquidTemplateTest.ItRendersAndCanBeParsedAsync.verified.txt (hunk 2 above)
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid.UnitTests/LiquidTemplateTest.ItRendersAndCanBeParsedAsync.verified.txt:1]: It's consumed by `Verify`, see https://github.com/microsoft/semantic-kernel/pull/6048#discussion_r1585139820

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs (hunk 1 above)
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:100]: see https://github.com/microsoft/semantic-kernel/pull/6048#issuecomment-2085957297

File: .gitignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply .gitignore:495]: @LittleLittleCloud can you configure the Verify function to use another directory for that purpose? 
Something like `%Temp%` for Windows OR `/tmp` or `/var/tmp` for Unix.

[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply .gitignore:495]: It needs to be applied to all other tests that uses Verify? I still think adding `*.received.txt` to `.gitignore` is the easiest and right pattern here.

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs [hunk 7]
```diff
@@ -16,22 +17,29 @@ namespace Microsoft.SemanticKernel.PromptTemplates.Liquid;
...
     private readonly Dictionary<string, object> _inputVariables;
 
     /// <summary>Initializes the <see cref="LiquidPromptTemplate"/>.</summary>
     /// <param name="config">Prompt template configuration</param>
-    /// <exception cref="ArgumentException"><see cref="PromptTemplateConfig.TemplateFormat"/> is not <see cref="LiquidPromptTemplateFactory.LiquidTemplateFormat"/>.</exception>
+    /// <param name="allowUnsafeContent">Whether to allow unsafe content in the template</param>
+    /// <exception cref="ArgumentException">throw if <see cref="PromptTemplateConfig.TemplateFormat"/> is not <see cref="LiquidPromptTemplateFactory.LiquidTemplateFormat"/></exception>
     /// <exception cref="ArgumentException">The template in <paramref name="config"/> could not be parsed.</exception>
-    public LiquidPromptTemplate(PromptTemplateConfig config)
+    public LiquidPromptTemplate(PromptTemplateConfig config, bool allowUnsafeContent = false)
```
Comment (markwallace-microsoft) [inline dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:34]: We should add logging, this can come later

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs [hunk 8]
```diff
@@ -16,22 +17,29 @@ namespace Microsoft.SemanticKernel.PromptTemplates.Liquid;
...
     {
```
Comment (markwallace-microsoft) [inline dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:35]: In other prompt templates we do this

```
Verify.NotNull(promptConfig, nameof(promptConfig));
Verify.NotNull(promptConfig.Template, nameof(promptConfig.Template));
```

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs (hunk 7 above)
Comment (LittleLittleCloud) [reply dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs:36]: Track in https://github.com/microsoft/semantic-kernel/issues/6033

File: .gitignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply .gitignore:495]: Agree, let's go for the simple and transparent approach. If someone needs to find the `*.received.txt` it will be easy and they won't have to search through the code to see where they've been written.

[the remark this answers is not in the document]
Comment (dmytrostruk) [reply .gitignore:495]: I would recommend going even more simpler and avoid using VerifyTests, and use just standard `Assert` calls from xUnit, as in our other tests.

[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply .gitignore]: Resolved by removing all newly added verify tests in this PR
